src/lib.rs

```diff
@@ -268,13 +269,16 @@ impl NixPath for [u8] {
     where
         F: FnOnce(&CStr) -> T,
     {
-        let mut buf = [0u8; PATH_MAX as usize];
-
         if self.len() >= PATH_MAX as usize {
             return Err(Errno::ENAMETOOLONG);
         }
 
+        #[allow(clippy::uninit_assumed_init)] // u8s do nothing when dropped
+        let mut buf: [u8; PATH_MAX as usize] = unsafe { MaybeUninit::uninit().assume_init() };
```

Comment:
The Rust documentation doesn't make it immediately clear that this is sound upon a quick read (and actually seems to imply it isn't), so I'll request more justification in your comment.

From https://doc.rust-lang.org/nightly/core/mem/union.MaybeUninit.html#safety:

> It is up to the caller to guarantee that the MaybeUninit<T> really is in an initialized state. Calling this (`.assume_init()`) when the content is not yet fully initialized causes immediate undefined behavior. 

It seems like it'd be required to call `assume_init()` _after_ populating `buf`, but that seems tricky to do without falling back to `copy_nonoverlapping`/`CStr::from_ptr`, which has performance implications for large paths (see #1583).

The nomicon has some [helpful things to say](https://doc.rust-lang.org/nomicon/unchecked-uninit.html), but even that example demonstrates the use of an array of `MaybeUninit` types followed by a `transmute`, rather than going directly to the desired type. 

TL;DR: This ultimately seems reasonable to me, but I'd like to see a more in-depth justification and also get @asomers thoughts.

Reply:
@RalfJung hope you don't mind the ping. :) Can you comment on if creating an `assume_init` array of `u8`s is legal so long as you write them before reading? And if not, what's the prefered way to let people perform write operations on a `MaybeUninit` slice when there's no API that supports it? I've actually had similar use cases before where there appears to be no solution besides marking the whole array as `assume_init`. Here's an example where I want to use a RNG's `fill_bytes` method:

```rust
let mut buf: [u8; 4096] = unsafe { MaybeUninit::uninit().assume_init() };
random.fill_bytes(&mut buf[0..used]);
file.write_all(&buf[0..used])?;
```

---

@rtzoeller Agreed, the `assume_init` requirements are confusing and seem to contradict each other. For example, they say "The `assume_init` is safe because the type we are claiming to have initialized here is a bunch of `MaybeUninit`s, which do not require initialization." A `u8` also doesn't require initialization, so does that make it too? Hopefully @<!-- -->RalfJung can answer. Otherwise, I'll continue digging into the commit history to see if I can find something.

Reply:
@SUPERCILEX can you try the following implementation and report back on performance? It much more directly mirrors the example usage of `MaybeUninit`, and in my (admittedly very synthetic) benchmarks seems to perform similarly to your original implementation (which surprised me).

```rust
fn with_nix_path<T, F>(&self, f: F) -> Result<T>
where
    F: FnOnce(&CStr) -> T,
{
    if self.len() >= PATH_MAX as usize {
        return Err(Errno::ENAMETOOLONG);
    }

    // MaybeUninit usage modelled after
    // https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#initializing-an-array-element-by-element
    let mut buf: [MaybeUninit<u8>; PATH_MAX as usize] =
        unsafe { MaybeUninit::uninit().assume_init() };
    // copy_from_slice() doesn't work for MaybeUninit types
    for (ptr, elt) in buf.iter_mut().zip(self) {
        ptr.write(*elt);
    }
    buf[self.len()].write(0); // NUL terminate the string

    let buf = unsafe { std::mem::transmute::<_, [u8; PATH_MAX as usize]>(buf) };
    match CStr::from_bytes_with_nul(&buf[..=self.len()]) {
        Ok(s) => Ok(f(s)),
        Err(_) => Err(Errno::EINVAL),
    }
}
```

Reply:
It's definitely not kosher to call assume_init() anything that isn't actually initialized.  This is the easy and safe way to implement that function without over-initializing, though it involves a heap allocation.  Is it fast enough?
```rust
    fn with_nix_path<T, F>(&self, f: F) -> Result<T>
            where F: FnOnce(&CStr) -> T {
        let mut v: Vec<u8> = Vec::from(self);
        v.push(0);
        match CStr::from_bytes_with_nul(&v) {
            Ok(s) => Ok(f(s)),
            Err(_) => Err(Errno::EINVAL)
        }
    }
```

Reply:
Ok, I've spent a gazillion hours reading what feels like everything under the sun about uninitialized memory. :grin: Here's my executive summary:

- Simply creating a variable with uninitialized memory for a type where not all bit patterns are valid is UB. `bool` is a good example: in the Rust abstract machine, `bool` can only ever be 0 or 1, so simply the act of declaring an uninitialized `bool` results in an invalid Rust program. This means the optimizer is free to assume you cannot possibly be in the current state (since a valid Rust program would never be) and therefore you can delete all the code after that variable declaration, insert a panic, etc.
- Creating a variable with uninitialized memory for a type where every bit pattern is valid is *not* UB. In LLVM IR, it will be represented as an `undef` (which is below a `poison` which is below UB). That said, it means that the optimizer can pick whatever value makes it happy and does *not* need to be consistent about its choices. Here's a slightly simplified version of a popular example:
  ```rust
  #![feature(bench_black_box)]

  use std::mem;

  fn always_returns_true(b: u8) -> bool {
      std::hint::black_box(b == 5) || std::hint::black_box(b != 5)
  }

  pub fn main() {
     let x: u8 = unsafe { mem::MaybeUninit::uninit().assume_init() };
     println!("{}", always_returns_true(x));
  }
  ```
  Run it: https://rust.godbolt.org/z/cPEa5dnxs

---

So where does that leave us? Well theoretically speaking, my code has one piece of UB (more accurately, unspecified behavior) and one correct `undef`. First, it creates a reference (`buf[..self.len()]`) that points to uninitialized memory, but references must always point to initialized memory. Therefore, not a valid Rust program. Second, we have a bunch of u8's that are `undef`. However, since we never read them before writing, we cannot observe their `undef` behavior so we don't care.

Now what about from a pragmatic perspective? Neither cases have been specified: https://github.com/rust-lang/unsafe-code-guidelines/blob/master/active_discussion/validity.md. So really, people are already using references "incorrectly" (even in the stdlib IIRC) so it's unlikely that this model will break anytime soon.

That said, **I figured out a way to rewrite this with no UB that I can tell**. We use an initialized raw pointer. Raw pointers can be whatever the heck we want, it's only when they are dereferenced that the data they point to must be valid. The only part I'm still a little iffy on is `buf.assume_init_ref()[..=self.len()]`: it creates a reference to an array that is partially initialized, then immediately creates another reference to the part of the array that is initialized. I think that's fine, but if we really wanted to go crazy, we could stick with raw pointers and create the slice directly: `slice::from_raw_parts(buf, self.len() + 1)` (though this is ~1M instructions slower for whatever reason).

With the new implementation, we get even faster which is cool:

```
20,704,687 (15.32%) 4,384,493 (14.41%) 2,744,994 (13.52%) 134 ( 1.23%) 13 ( 0.04%) 265 ( 1.72%) 8 ( 0.15%) 0          45 ( 0.37%) 2,441,283 (12.30%) 1,160 ( 0.23%) 609,972 (26.51%)  4 ( 0.04%)  => /home/asaveau/Desktop/nix/src/lib.rs:<[u8] as nix::NixPath>::with_nix_path (152,493x)
```

@rtzoeller your implementation benchmarks similarly in terms of time taken, but suffers significantly on the instruction count side:
```
180,403,130 (61.18%) 56,855,699 (68.57%) 39,502,833 (69.22%) 590 ( 4.24%) 5,281 (12.19%) 7,548 (32.57%) 12 ( 0.22%) 701 (14.96%) 1,069 ( 8.01%) 18,761,972 (51.87%) 305,759 (37.73%) 6,101,922 (78.29%) 1,513 (14.69%)  => /home/asaveau/Desktop/nix/src/lib.rs:<[u8] as nix::NixPath>::with_nix_path (152,496x)
```

@asomers I'm *very strongly* opposed to creating an allocation per syscall. The whole point of this PR is to be completely zero-cost. Interestingly enough, allocating is not the worst on the instruction count front:
```
48,197,762 (29.65%) 13,229,886 (33.70%) 8,234,742 (31.94%) 135 ( 1.23%) 34 ( 0.11%) 9 ( 0.06%) 7 ( 0.13%) 0          0          6,253,608 (26.44%) 2,324 ( 0.46%) 1,524,930 (47.41%) 15 ( 0.20%)  => /home/asaveau/Desktop/nix/src/lib.rs:<[u8] as nix::NixPath>::with_nix_path (152,493x)
```
Though when I ran the benchmarks from #1583, it performed significantly worse.

Reply:
Mmmmmm, we need to bump the MSRV to `1.55.0` to use `assume_init_ref`. Note that we can't use `assume_init` unfortunately because that makes a copy of the array.

Having to resort to raw pointers is kind of a bummer since it's uglier and produces a teeny bump in instruction count:

```
21,468,780 (15.80%) 4,689,831 (15.26%) 2,897,707 (14.16%) 475 ( 3.12%) 19 ( 0.06%) 320 ( 1.99%) 8 ( 0.15%) 0          48 ( 0.39%) 2,593,974 (12.97%) 1,351 ( 0.26%) 762,520 (31.06%) 606 ( 6.21%)  => /home/asaveau/Desktop/nix/src/lib.rs:<[u8] as nix::NixPath>::with_nix_path (152,504x)
```

Reply:
@asomers forgot to mention that I also opened a PR that removes the `PATH_MAX` restriction with a heap allocation so we can avoid probe frames: https://github.com/nix-rust/nix/pull/1656

Reply:
> Creating a variable with uninitialized memory for a type where every bit pattern is valid is not UB.

That is not correct -- the [Rust Reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) says this is UB, too. Indeed, `undef` is *not* just "some unknown bit pattern", it is a value that is [clearly distinct from any single 'regular' bit pattern](https://www.ralfj.de/blog/2019/07/14/uninit.html), and `u8` is well-defined for any regular bit pattern but not for `undef`.

The current `Write` trait does not adequately support writing into an uninitialized buffer; [this RFC](https://rust-lang.github.io/rfcs/2930-read-buf.html) describes the planned changes to make it compatible. Other, similar APIs should probably use the same pattern.

Reply:
> That is not correct -- the Rust Reference says this is UB, too.

Isn't that still unspecified though? Or at least, I thought the validity discussions were about deciding whether or not it should be considered UB for Rust. From the LLVM perspective, an `undef` u8 *is* defined (as far as I understand). For example, `undefu8 | 3 = uuuuuu11`. Those `u`s can be whatever and change whenever, but we know that no matter what LLVM chooses for undef, the lower two bits must be 1. That's what I got from [these slides](https://llvm.org/devmtg/2020-09/slides/Lee-UndefPoison.pdf).

> this RFC describes the planned changes to make it compatible.

That's freakin' awesome! Glad to see progress. :)

Reply:
Ohhhh, that makes perfect sense. Thank you for all the feedback!